#! /bin/sh /usr/share/dpatch/dpatch-run
## 20_API-0.dpatch by Sebastien Delafond <seb@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Honor size=0 FUSE API (Closes: #524655).

@DPATCH@
diff -urNad fuse-python-0.2~/fuseparts/_fusemodule.c fuse-python-0.2/fuseparts/_fusemodule.c
--- fuse-python-0.2~/fuseparts/_fusemodule.c	2007-06-18 07:20:09.000000000 -0700
+++ fuse-python-0.2/fuseparts/_fusemodule.c	2009-04-18 13:25:01.000000000 -0700
@@ -671,8 +671,22 @@
 #endif
 
 	if(PyString_Check(v)) {
-		if(PyString_Size(v) > size)
+        /* size zero can be passed into these calls  to return the current size of
+         * the named extended attribute
+         */
+        if (size == 0) {
+		    ret = PyString_Size(v);
+			goto OUT_DECREF;
+        } 
+
+        /* If the size of the value buffer is too small to hold the result,  errno
+         * is set to ERANGE.
+         */
+		if (PyString_Size(v) > size) {
+            ret = -ERANGE;
 			goto OUT_DECREF;
+        }
+
 		memcpy(value, PyString_AsString(v), PyString_Size(v));
 		ret = PyString_Size(v);
 	}
